src/main/java/org/wildfly/security/http/impl/BearerTokenAuthenticationMechanism.java

```diff
@@ -81,41 +82,35 @@ public String getMechanismName() {
         List<String> authorizationValues = request.getRequestHeaderValues(HttpConstants.AUTHORIZATION);
 
-        if (authorizationValues == null || authorizationValues.isEmpty()) {
-            request.authenticationFailed("Bearer token required", this::unauthorizedResponse);
-            return;
-        } else if (authorizationValues.size() > 1) {
-            request.authenticationFailed("Multiple Authorization headers found", response -> response.setStatusCode(400));
-            return;
-        }
-
-        String authorizationValue = authorizationValues.get(0);
-        Matcher matcher = BEARER_TOKEN_PATTERN.matcher(authorizationValue);
-
-        if (!matcher.matches()) {
-            request.authenticationFailed("Authorization is not Bearer", response -> response.setStatusCode(400));
-            return;
-        }
-
-        BearerTokenEvidence tokenEvidence = new BearerTokenEvidence(matcher.group(1));
-        EvidenceVerifyCallback verifyCallback = new EvidenceVerifyCallback(tokenEvidence);
-
-        handleCallback(verifyCallback);
-
-        if (verifyCallback.isVerified()) {
-            AuthorizeCallback authorizeCallback = new AuthorizeCallback(null, null);
-
-            handleCallback(authorizeCallback);
-
-            if (authorizeCallback.isAuthorized()) {
-                handleCallback(new IdentityCredentialCallback(new BearerTokenCredential(tokenEvidence.getToken()), true));
-                handleCallback(AuthenticationCompleteCallback.SUCCEEDED);
-                request.authenticationComplete();
-                return;
+        if (authorizationValues != null) {
+            Matcher matcher;
+            for (String current : authorizationValues) {
+                if ((matcher = BEARER_TOKEN_PATTERN.matcher(current)).matches()) {
+                    BearerTokenEvidence tokenEvidence = new BearerTokenEvidence(matcher.group(1));
+                    EvidenceVerifyCallback verifyCallback = new EvidenceVerifyCallback(tokenEvidence);
+
+                    handleCallback(verifyCallback);
+
+                    if (verifyCallback.isVerified()) {
+                        AuthorizeCallback authorizeCallback = new AuthorizeCallback(null, null);
+
+                        handleCallback(authorizeCallback);
+
+                        if (authorizeCallback.isAuthorized()) {
+                            httpBearer.debugf("Token authentication successful.");
+                            handleCallback(new IdentityCredentialCallback(new BearerTokenCredential(tokenEvidence.getToken()), true));
+                            handleCallback(AuthenticationCompleteCallback.SUCCEEDED);
+                            request.authenticationComplete();
+                            return;
+                        }
+                    }
+                    httpBearer.debugf("Token authentication failed.");
+                    request.authenticationFailed("Invalid bearer token", response -> response.setStatusCode(FORBIDDEN));
```

Comment:
Here needs to be added `return` to prevent following tokens verification!
Otherwise attacker is able to try multiple tokens in one request.

Reply:
It is good to allow multiple authorization headers as introduced in this, but only one should be used for one mechanism.